### PR TITLE
refactor: move/rename internal emoji methods

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

Moves `PartialEmoji._emoji_from_name_id` to state, updates typing of existing method to be more accurate (mainly for #928), and simplifies emoji handling for welcomescreen channels.

The two methods are arguably very similar, but there are still some slight differences, mostly due to workarounds for weird API things; merging them into one isn't feasible :/

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `pdm lint`
    - [x] I have type-checked the code by running `pdm pyright`
- [ ] This PR fixes an issue
- [ ] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: disnake/reaction.py [hunk 1]
```diff
@@ -64,7 +64,8 @@ def __init__(
         emoji: Optional[Union[PartialEmoji, Emoji, str]] = None,
     ) -> None:
         self.message: Message = message
-        self.emoji: Union[PartialEmoji, Emoji, str] = emoji or message._state.get_reaction_emoji(
+        # _get_emoji_from_data won't return None
+        self.emoji: Union[PartialEmoji, Emoji, str] = emoji or message._state._get_emoji_from_data(  # type: ignore
```
Comment (onerandomusername) [inline disnake/reaction.py:68]: but why do we need a type ignore?

File: disnake/state.py [hunk 2]
```diff
@@ -1911,18 +1911,66 @@ def _get_reaction_user(
             return channel.guild.get_member(user_id)
         return self.get_user(user_id)
 
-    def get_reaction_emoji(self, data) -> Union[Emoji, PartialEmoji]:
-        emoji_id = utils._get_as_snowflake(data, "id")
+    # methods to handle all sorts of different emoji formats
+
+    def _get_emoji_from_data(
+        self, data: PartialEmojiPayload
+    ) -> Optional[Union[str, Emoji, PartialEmoji]]:
```
Comment (onerandomusername) [inline disnake/state.py:1918]: how can this return None? I don't see a codepath that allows that.

**Review by onerandomusername — CHANGES_REQUESTED** [review]

File: disnake/reaction.py (hunk 1 above)
Comment (shiftinv) [reply disnake/reaction.py:68]: reaction emojis always have at least an ID or name, so this always returns *something*

File: disnake/state.py (hunk 2 above)
Comment (shiftinv) [reply disnake/state.py:1918]: it's not immediately apparent, but `data["name"]` is nullable

**Review by onerandomusername — APPROVED** [review]

File: disnake/welcome_screen.py [hunk 3]
```diff
@@ -74,17 +74,10 @@ def _from_data(
...
-        emoji_name = data.get("emoji_name") or None
-        emoji = None
-        if emoji_name:
-            emojis: Optional[Tuple[Emoji]]
-            if emojis := getattr(guild, "emojis", None):
-                emoji = utils.get(emojis, id=emoji_id, name=emoji_name)
-            if not emoji:
-                emoji = PartialEmoji.with_state(state, name=emoji_name, id=emoji_id)
+        emoji_name = data.get("emoji_name")
+        emoji = state._get_emoji_from_fields(name=emoji_name, id=emoji_id)
```
Comment (onerandomusername) [inline disnake/welcome_screen.py:80]: style inconsistency between this and the above

File: disnake/channel.py [hunk 4]
```diff
@@ -3050,8 +3050,8 @@ def default_reaction(self) -> Optional[Union[Emoji, PartialEmoji]]:
 
         .. versionadded:: 2.6
         """
-        return PartialEmoji._emoji_from_name_id(
-            self._default_reaction_emoji_name, self._default_reaction_emoji_id, state=self._state
+        return self._state._get_emoji_from_fields(
+            name=self._default_reaction_emoji_name, id=self._default_reaction_emoji_id
```
Comment (onerandomusername) [inline disnake/channel.py:3054]: see below comment

File: disnake/welcome_screen.py (hunk 3 above)
Comment (shiftinv) [reply disnake/welcome_screen.py]: https://github.com/DisnakeDev/disnake/pull/980/commits/3819f3eed4cc25fc8a979109e02d933263291d4f

**Pull request merged**